### DISCUSSION TITLE
Ideas for Improving ProcessWrapper

### DISCUSSION THREAD

**Discussion by wonderbird** [wonderbird/TestProcessWrapper#4] [Ideas]
The questions and suggestions collected in this thread have risen (and rise) from a small group of developers talking about code every now and then.

**Comment by wonderbird** [comment]
DONE: Replace the conditional in `InvokeKillSystemCall` with a command or strategy pattern

**Comment by wonderbird** [comment]
Ensure that the parent process also ends, even when the BDD tests fail

**Comment by wonderbird** [comment]
DONE: Actually, I would go for Factory pattern to create Process object. but not .net process object but my domain Process object. lets say `IProcess` interface

```
WindowsProcess : IProcess
OS2Process : IProcess
IOProcess : IProcess
```

and so on
 
`IProcess` contains required methods of course 
 
```
class Factory{
  public static IProcess Create(){} 
}
```

Basically, Factory gives me the Process I need, so I don't have to think about the process itself. Also it supports separation of concerns and also it's open for extension but closed for modifications. Basicly, u can support SOLID
 
Good part is, inside factory, now u are using Process object but in the future you can switch to Powershell object itself to manipulate process.

**Comment by alex-held** [comment]
DONE. This is now a readiness check (Pattern) located in the class ProcessIdReader.

``` cs
     private void ParseStartupMessage(string startupMessage)
        {
            if (_dotnetHostProcessId.HasValue || !startupMessage.Contains("Process ID"))
            {
                return;
            }

            var processIdStartIndex = startupMessage.IndexOf("Process ID", StringComparison.Ordinal);
            var newLineAfterProcessIdIndex =
                startupMessage.IndexOf("\n", processIdStartIndex, StringComparison.Ordinal);
            var processIdNumberOfDigits = newLineAfterProcessIdIndex - processIdStartIndex - 10;
            var processIdString = startupMessage.Substring(processIdStartIndex + 10, processIdNumberOfDigits);
            _dotnetHostProcessId = int.Parse(processIdString, NumberStyles.Integer, CultureInfo.InvariantCulture);

            // throw exception to prevent unwanted program behavior later on
            TestOutputHelper?.WriteLine($"Process ID: {_dotnetHostProcessId.Value}");
        }

```
